keps/sig-auth/3766-referential-authorization/README.md

```diff
@@ -0,0 +1,1212 @@
+  specific `To` Kinds.
+* Corollary for future work, define how controllers interact. Is it a problem if
+  multiple controllers consume the same ReferenceGrant?
+* Similar to the Gateway API version of ReferenceGrant, controllers will
+  indicate status on the _referring_ resources, not on ReferenceGrant itself.
+* Clarify that the expected operating model for implementations expects them to
+  have broad, read access to both ReferenceGrant and the specific `To` Kinds they
+  support, and then self-limit to only _use_ the relevant ones.
+* Determine if strings used in `For` are completely separate from verbs.
+* Decide if supporting verbs such as `[create, update, delete]` make sense.
+* Decide if supporting virtual verbs such as `[forward, scale, dance]` make sense.
+  We already have `For` which has a similar function.
+* Decide whether to expose/hide the API based on the API server feature flag
+  and/or authorizer config.
+* Decide how cross-namespace ReferenceGrants join target resources to their
+  originating reference. One way of accomplishing this is to extract the
+  namespace alongside the name of a target reference when reconciling 
+  ClusterReferenceGrants, then pair those with a matching ReferenceGrant. A 
+  trade-off of this approach is that it makes ReferenceGrants
+  dependent on ClusterReferenceGrants. The alternative is that the subject of a ClusterReferenceConsumer is authorized for the granted object regardless of 
+  whether there is actually a cross-namespace reference.
+* Decide whether to formally add `*` as a special value for Namespace, to mean
+  "all namespaces".
+* Decide whether to formally add `*` as a special value for classNames, to mean
+  "all classNames". Should access to a class-partitioned API be open or closed 
+  for a ClusterReferenceConsumer that lists no classNames? Software authors, API
+  owners, and cluster admins all own different pieces of the config here. An 
+  optional field that fails open can easily be missed by a cluster admin. An  
+  opt-in wildcard can become canonical in practice.
+* Adding classNames burdens cluster admins with synchronizing their classNames 
+  between their IngressClasses, GatewayClasses, StorageClasses, etc.
+  IngressClass and GatewayClass already have a `spec.controller` and
+  `spec.controllerName` field. StorageClass has a `provisioner` field.
+  There is an opportunity here to have ClusterReferenceConsumers select into
+  these class types by their controller/provisioner name.
+  It requires API owners to define how to extract which of classNames pair up 
+  with which controller/provisioner names.
+  The benefit is that ClusterReferenceConsumers can specify a more canonical
+  controllerName field, leaving all of the declarations of classNames contained
+  within their specialized IngressClass/GatewayClass/StorageClass API.
+  This extra layer of indirection reduces config toil and duplication, but is
+  arguably more incomprehensible.
```

Comment:
I've currently proposed just extracting classNames from referencing objects and matching them to consumers.
Then cluster-admins have to ensure classNames are correct for a consumer if they change them.
( maybe this get's templated by a helm-chart/operator/packageinstall ? )

I'm curious for feedback for this other idea of having another API definition mapping `spec.controller` and `provisioner` to a `ClusterReferenceConsumer.controllerName` so we can match GatewayClasses and IngressClasses to their Subject. It could live inside of ClusterReferenceGrant, or it could be it's own thing like ClassControllerMap:

```
kind: ClassControllerMap
apiVersion: reference.authorization.k8s.io/v1alpha1
metadata:
  name: gatewayclass
spec:
  versions:
  - version: v1
    group: gateway.networking.k8s.io/v1
    resource: gatewayclass
    controllerPath: .spec.controllerName
```